### ISSUE TITLE
New linter for comment spacing

### ISSUE THREAD

**Issue by MichaelChirico** [r-lib/lintr#1132]
`styler` enforces comment spacing:

```r
styler::style_text("a #b")
# a # b
```

We could throw a lint calling out this rule as well.

This is part of the style guide, so it should be a default linter:

https://style.tidyverse.org/syntax.html#comments

> Each line of a comment should begin with the comment symbol and a single space: #

**Comment by AshesITR** [comment]
I think we should be a little bit more lenient than requiring exactly one space by default, though.

Using more than 1 space can be useful to emphasize, e.g. an equation.
```
# The foo can be determined by:
#    (bar + baz) ^ 2 - boo
```

Also, I prefer `#> 1 + 1` to `# > 1 + 1` as a workaround to commented code lints.
WDYT about requiring at least one space and optionally a, possibly repeated, preceding non-alphanumeric?

This would also allow flexible "heading" comments like
`#### Heading ####`.

**Comment by MichaelChirico** [comment]
I was thinking to let the allowed chars be configurable. that way it covers roxygen comments, but also any other meta commenting systems that might be out there. by default we'd allow ' # and > to come right after #.

good point about hanging indent comments... that's definitely my own personal style. we could (1) only enforce the one-space rule on the first line of a block of comments (2) have a strict mode (like for assignment_spaces_linter) that toggles whether exactly-one or just non-zero spaces is enforced or (3) always allows a number>0 of spaces

**Comment by AshesITR** [comment]
I like (1) with an option to activate (3).
Good thinking of the roxygen style.

A block would be defined as `#` starting on the same column in subsequent lines, right?

**Comment by MichaelChirico** [comment]
yep so

```
foo() # c1
double_foo() #c2
```

will lint on c2

**Comment by AshesITR** [comment]
How do we want to deal with repetitions of allowed symbols?
I can generally see them used for heading styles, e.g. `#----- a -----` and `##### a #####`.

So block comment starts should match `^#[#'>-]*\ [^\ ]` and continuation lines should match `^#[#'>-]*\ +[^\ ]` by default, right?

**Comment by MichaelChirico** [comment]
that sounds right. my first reaction was that subsequent lines should match the same repetition but that would preclude decorative comments like

```
############
# comment #
############
```

(although, it would be fine if the decorative comment starts with a space...)

**Comment by MichaelChirico** [comment]
this should also lint "Emacs" / Bioconductor style (by default, with option of course):

```r
## used for indentation rules
```

https://contributions.bioconductor.org/r-code.html#rcode-comments

**Comment by ahl27** [comment]
good point, I didn't even know this was a difference from the tidyverse style until michael pointed it out to me lol

I filed a quick feature request to add general bioc support as well--it shouldn't be a terribly difficult feature to implement (https://github.com/r-lib/lintr/issues/3027).
